### ISSUE TITLE
dice_loss , valid_mask is not applied on the denominator

### ISSUE THREAD

**Issue by syzygyfy** [open-mmlab/mmsegmentation#1088]
In dince_loss.py
https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/losses/dice_loss.py, Line 45
why is that valid_mask  is not applied on the denominator ?
`den = torch.sum(pred.pow(exponent) + target.pow(exponent), dim=1) + smooth`

In this implementation it is applied , 
https://github.com/LikeLy-Journey/SegmenTron/blob/master/segmentron/solver/loss.py, Line 306
`den = torch.sum((predict.pow(self.p) + target.pow(self.p)) * valid_mask, dim=1) + self.smooth`

**Comment by MengzhangLI** [comment]
Hi, sorry for late reply. 

I think it is OK because the value of `den` is a constant which would not affect back propagation. So calculation of `valid_mask` in `den` is unnecessary. What's more, keeping the same `den` between with and without `igore_index` situation would directly show change of loss.

**Comment by syzygyfy** [comment]
Hi, Thank you for the response.
consider following example:
In this case i am calculating loss only on class_id=1
```
import numpy as np
import torch
x = np.zeros((1,2,5,5),dtype=np.uint8)
y  = np.zeros((1,5,5),dtype=np.uint8)
x[:,1:,0:2,0:2]=12
y[:,0:2,0:2]=1
x = torch.Tensor(x)
y = torch.Tensor(y)
l = DiceLoss(ignore_index=0)
print (l(x,y))
```
with mask, it produces following results:
```
tensor(0.)
```
and without valid_mask ( current implementation ) following results:

```
tensor(0.1842)
```

Although, with default value of ignore_index , it has no effect, so it shouldn't be a problem.

**Comment by syzygyfy** [comment]
I am closing this as there is no further activity.
Thank you
